backend-services/user-verification-service/src/main/java/org/cardano/foundation/voting/service/discord/DefaultDiscordUserVerificationService.java

```diff
@@ -119,58 +125,98 @@ public Either<Problem, DiscordStartVerificationResponse> startVerification(Strin
+            default:
+                return Either.left(Problem.builder()
+                        .withTitle("UNSUPPORTED_WALLET_TYPE")
+                        .withDetail("The specified wallet type is not supported.")
+                        .withStatus(BAD_REQUEST)
+                        .build());
+        }
+    }
 
-        if (event.finished()) {
-            log.warn("Event already finished:{}", eventId);
+    private Either<Problem, EventSummary> validateEvent(String eventId) {
+        Either<Problem, Optional<EventSummary>> eventDetailsResult = chainFollowerClient.findEventById(eventId);
+
+        if (eventDetailsResult.isLeft()) {
+            return Either.left(eventDetailsResult.getLeft());
+        }
+
+        return eventDetailsResult.get()
+                .map(Either::<Problem, EventSummary>right)
+                .orElseGet(() -> Either.left(Problem.builder()
+                        .withTitle("EVENT_NOT_FOUND")
+                        .withDetail("Event not found, eventId: " + eventId)
+                        .withStatus(BAD_REQUEST)
+                        .build()));
+    }
 
+    private Either<Problem, IsVerifiedResponse> handleCardanoVerification(DiscordCheckVerificationRequest request, String eventId, String walletId) {
+
+        String signature = request.getCoseSignature().orElse(null);
+
+        if (signature == null) {
             return Either.left(Problem.builder()
-                    .withTitle("EVENT_ALREADY_FINISHED")
-                    .withDetail("Event already finished, eventId:" + eventId)
+                    .withTitle("MISSING_SIGNATURE")
+                    .withDetail("Missing signature.")
                     .withStatus(BAD_REQUEST)
                     .build());
         }
 
-        var requestStakeAddress = checkVerificationRequest.getStakeAddress();
-        var requestSecret = checkVerificationRequest.getSecret();
-
-        var coseSignature = checkVerificationRequest.getCoseSignature();
-        var cosePublicKey = checkVerificationRequest.getCosePublicKey();
+        String publicKey = request.getCosePublicKey().orElse(null);
 
-        var cip30Verifier = new CIP30Verifier(coseSignature, cosePublicKey);
-        var cip30VerificationResult = cip30Verifier.verify();
-
-        if (!cip30VerificationResult.isValid()) {
+        if (publicKey == null) {
             return Either.left(Problem.builder()
-                    .withTitle("INVALID_CIP-30-SIGNATURE")
-                    .withDetail("Invalid CIP-30 signature")
+                    .withTitle("MISSING_PUBLIC_KEY")
+                    .withDetail("Missing public key.")
                     .withStatus(BAD_REQUEST)
-                    .build()
-            );
+                    .build());
+        }
+
+        // Verify signature specific to Cardano wallets
+        Either<Problem, Tuple2<String, Optional<String>>> verificationResult = verifySignature(
```

Comment:
General rule: It all depends but generally speaking one tries to avoid usage of too many tuples. One can simply create a business / domain specific class. It is however, generally ok to use tuples within the same class as context to understand is smaller.

Reply:
Okey, got it.
I created this class:

```java
package org.cardano.foundation.voting.service.common;
import java.util.Optional;

public class VerificationResult {
    private final String message;
    private final Optional<String> address;

    public VerificationResult(String message, Optional<String> address) {
        this.message = message;
        this.address = address;
    }

    public String getMessage() {
        return message;
    }

    public Optional<String> getAddress() {
        return address;
    }
}
```